### PR TITLE
[v7r2] HtCondorCE: request MP slots

### PR DESCRIPTION
**Pull request by fstagni**
BEGINRELEASENOTES

*Resources
FIX: HtCondorCE: corrected ClassAd for requesting MP slots

ENDRELEASENOTES

### CODE REVIEW

**Comment by marianne013** [comment]
I quickly dumped the original 2 line fix on my server, but that didn't have the anticipated effect.
FYI: @sfayer

**Comment by marianne013** [comment]
@fstagni When I looked at this on my test server, the issue was far more fundamental: It looked  like the number of cores requested ("processors") never reached the the code that makes the condor sub file, hence it would always request 1. (Found by crude print statement.)
I tried to verify this on the certification server, but that seems to have gone into a complete sulk and none of my jobs (not even the single cores ones) get submitted at the moment. Before we go down the condor version rabbit hole, would it be possible to verify that the number of processors is actually seen by the HTCondorCE code in v7r2 and higher ? Maybe a question for @andresailer ?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 1]
```diff
@@ -239,7 +239,8 @@ def __writeSub(self, executable, nJobs, location, processors):
 initialdir = %(initialDir)s
 grid_resource = condor %(ceName)s %(ceName)s:9619
 transfer_output_files = ""
-+xcount = %(processors)s
+machine_count = 1
```
Comment (aldbr) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:242]: `machine_count` seems to depend on the `parallel` universe according to the documentation: https://htcondor.readthedocs.io/en/v8_8/man-pages/condor_submit.html

It is described in the *COMMANDS FOR PARALLEL, JAVA, and SCHEDULER UNIVERSES* section.
Using `vanilla` or `grid` universe implies exploiting `n` cpus - specified by `request_cpus` - from one machine per allocation I guess.

**Comment by aldbr** [comment]
> @fstagni When I looked at this on my test server, the issue was far more fundamental: It looked like the number of cores requested ("processors") never reached the the code that makes the condor sub file, hence it would always request 1. (Found by crude print statement.)

I guess there is a missing line in `HTCondorComputingElement.__writeSub` to get the number of processors from the `ceParameters` such as:

``` python
def __writeSub(self, executable, nJobs, location, processors):
...
processors = processors if processors > 1 else self.ceParameters.get('NumberOfProcessors', processors)
...
```

This should allow `HTCondorComputingElement` to get the number of processors defined in the CE configuration.

**Comment by fstagni** [comment]
Thanks! I made the modifications.

**Comment by andresailer** [comment]
I think the real question is why the SiteDirector doesn't have the processors argument in submitjob
https://github.com/DIRACGrid/DIRAC/blob/d4f7d3e4a3312588b402e77267a102987b07d163/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L837
The MultiProcessorSiteDirector used to have it
https://github.com/DIRACGrid/DIRAC/blob/1e27806a272066be3053114fcaf9648dd93af857/WorkloadManagementSystem/Agent/MultiProcessorSiteDirector.py#L336

**Comment by marianne013** [comment]
@andresailer That's why it used to work, I guess...

**Comment by fstagni** [comment]
The SiteDirector should not have the processors argument in `submitJob`, as this is an information handled at the ComputingElement level.
There is, it seems, some old code still lying around. I see what I can do to harmonize things.

**Comment by marianne013** [comment]
I don't quite understand how this is meant to work if I have users submitting 1 core and 8 core jobs to the same CE. Surely at some point it must get the values from the JDL.

**Comment by fstagni** [comment]
One thing is the Job requirement (the JDL), and another thing is the "size" of the Pilot. The requirements of the first do not influence the second (in other words: we don't submit pilots based on what are the requirements of the jobs in the queue.

To your favorite _CE.some.where.org/its_queue_ the SiteDirector will submit all identical pilots, based on what is the CE/queue description. If that CE/queue wants a 8-processors pilots, then we'll submit 8-processors pilots with **PoolCE**, which is capable of matching jobs of different "sizes" until full.

**Comment by marianne013** [comment]
Sigh. Let me tell you the story of a VO that ran 64 core jobs at RAL and the poor little nagios job. Now we have to code ourselves out of an LHCb shaped hole again.

**Comment by atsareg** [comment]
Pilots concept is not very good indeed in specifying very precise requirements to the slot on a WN. So, we can not translate to the pilot requirements some specific requirements of jobs, otherwise these would be pilots for particular jobs, which is equivalent to sending jobs directly loosing all the advantages of the pilots. But in the case above (a mixture of 64 core and single-core jobs in the same CE), the usual thing is to define two queues identical but for the NumberOfProcessors parameter. So, one queue will have NumberOfProcessors=64 and RequiredTag=MultiProcessor, the other queue will not have any of those, meaning single processor queue. In this case for 64-core and single-core jobs different slots will be reserved by the pilots if we do not want to mix such jobs. But in any case, the original problem was just triggered by a bug that is found and the PoolCE should be able to cope with all the cases. Just that sometimes a single single-core job will be executed in a fat 64-core slot. But this should be rare as we usually do not submit jobs one by one. If this is the case, the scenario described above can help.

**Comment by marianne013** [comment]
@atsareg So I should define two queues in DIRAC (one with e.g. 8 processors and one with 1) and have them both point to the same queue at the site ? If so, could someone please fix the certification server so I can try it out there ?

**Comment by fstagni** [comment]
> @atsareg So I should define two queues in DIRAC (one with e.g. 8 processors and one with 1) and have them both point to the same queue at the site ?

Exactly. There's not necessary a 1-to-1 relation between what's defined in DIRAC and what's defined "in real", DIRAC will always apply a "vision" on top of it. E.g. nothing prevents you from putting each and every CE/queue out there in a single DIRAC site (that would not be very good accounting-wise).

> If so, could someone please fix the certification server so I can try it out there ?

Which site/ce/queue?

**Comment by marianne013** [comment]
@fstagni 
The certification server currently doesn't submit pilots, due to (I think) problems in the OptimizationMind. I was going to use Imperial for testing as I can see the jobs and their properties arriving on the other end and besides, I would only screw up my own site.

**Review by atsareg — APPROVED** [review]

**Comment by marianne013** [comment]
@fstagni @atsareg 
While I can confirm this works in the sense that I can now submit 8 core pilots (I pulled Federico's v7r2 branch on one of my test servers), I don't think the one core vs multicore job distribution quite works yet.

**Comment by atsareg** [comment]
I missed something probably. But what is the problem with the one core vs multicore job distribution ? You can define queues which are suitable for ones and for others (or for both). So, everyone will get served - isn't ?

**Comment by marianne013** [comment]
@atsareg Does the single core queue also have to have LocalCEType = Pool or not (I thought not, but I might be wrong).
If you look at ceprod03 on the certification server, that's how I understood the configuration to work.

**Comment by atsareg** [comment]
Pool should be a universal solution although somewhat an overkill for the single-core queues. Your definition of the ceprod03 looks almost fine to me. The Tag = MultiProcessor does not make sense, it will be added anyway. I think you meant RequiredTag = MultiProcessor. In this case the queue will only receive jobs with more than 1 processor requirement. RequiredTag is the CE/Queue requirement to the jobs. This is relevant since you have another queue for single processor jobs only. Note that if a job requires NumberOfProcessors = N (N>1), then the MultiProcessor tag will be added automatically to the job requirements and the CE/Queue will have a possibility to request jobs only with this tag.

MaxWaitingJobs is probably too much.

**Comment by marianne013** [comment]
@atsareg The "Tag = MultiProcessor" is from the example Federico gave me on the forum. If I put in "RequiredTag = MultiProcessor" would  I then have to ask my users to define
NumberOfProcessor = 8 && Tag = MultiProcessor or is NumberOfProcessor = 8  sufficient ? 
I can take the Pool CE out from the single core queues

**Comment by atsareg** [comment]
NumberOfProcessors = 8 in the jobs will be sufficient, in this case Tag = MultiProcessor will be added by the WMS to the job description to be used in the (dual) matching mechanism.

**Comment by fstagni** [comment]
If you put "RequiredTag = MultiProcessor" then the CE/queue will ONLY accept MP jobs. 
Your users should not fiddle with tags, they should rather define their jobs with

```
from DIRAC.Interfaces.API.Job import Job
Job().setNumberOfProcessors()  <-- check this one.
```

**Pull request merged**